[MME] Implicit Network-initiated Deregistration

jmasterfunk84 opened the pull request:
This pull request creates three new configurable EMM timers as defined in 24.301.  Their values can be set in mme.yaml.  If the value is set to zero, the timer is disabled.  Their default values are configured as below:
* mobile_reachable (56 minutes) &mdash; This value should be 4 minutes greater than T3412.
* implicit_detach (4 minutes) &mdash; Upon expiry, the UE will be implicitly detached.
* purge_ue (24 hours) &mdash; The UE Context will be removed from the MME, and the HSS informed if applicable.

When the Implicit Detach timer is running, paging to the UE will be unavailable.  Resources which may trigger paging will be informed of the unavailability of paging.

An issue was also uncovered where a detach could arrive at the MME from an unknown UE.  This would create a UE context that could not be cleared by the purge_ue timer.  The detach from the unknown UE will now receive `UE_IDENTITY_CANNOT_BE_DERIVED_BY_THE_NETWORK`, and be transitioned into the EMM exception state.  Additional debug statements added to the EMM state machine.

acetcom commented:
@jmasterfunk84 

1. Does the standard refer to anything about the Purge Timer? Please provide a reason(or rerference) for implementing this.

2. You made TIMER configurable. And also you created mme_s1ap_page_if_attached() function. However, AMF didn't do that. Open5GS does not allow the two to be written differently. It is recommended that you follow the method used by AMF whenever possible, and if not, you should explain why and fix both.

Thanks a lot!
Sukchan

jmasterfunk84 commented:
@acetcom,

1. Purge UE is mentioned in 23.401, 5.3.9.3.  "The MME may, as an implementation option, delete the subscription data and MM context of an UE immediately after the implicit or explicit detach of the UE. Alternatively the MME may keep for some time the subscription data and the MM context of the detached UE, so that the data can be reused at a later attach without
accessing the HSS."  The commercial MMEs I have experience with include a Purge Timer.

2. I created mme_s1ap_page_if_attached() as there were instance of s1ap_send_paging.  I felt that the added function would reduce the amount of duplicated code where we check if the implicit_detach timer is running.
May I also make the AMF timers configurable?

jmasterfunk84 commented:
I'll also note that 23.502 4.5.3 seems to indicate that the immediate removal of the subscriber data and MM context is optional.  I don't have any commercial AMF experience, however, I think this clause indicates the AMF could also support a purge timer as well.

acetcom commented:
@jmasterfunk84 

1. Purge  Timer
It means that the Purge UE is optionally done when it is an Implicit/Explicit Detach/Deregistration UE. Purge Timer is only metioned as an **alternative** in 23.401, and not even mentioned in 23.502. Therefore, I don't think Purge Timer should be implemented in Open5GS.

2. Configuration Timer
Open5GS strives to minimize configuration files. Because it can confuse users. Therefore, AMF is implemented as a relative value for t3512.

3. mme_s1ap_page_if_attached()
This function could be efficient, but it makes source code less readable.

From what I've reviewed so far, it seems like it would be better to do it the way AMF implements it.

Please let me know if you have any different idea.

Thanks a lot!
Sukchan

jmasterfunk84 commented:
Hi @acetcom,
1. I personally really like the Purge Timer as it reduces the HSS signaling for a subscriber that might spend long periods of time out of service.  However, I agree it's not strongly specified for the AMF.  It has been removed.
2. I've removed the configuration for those 2 remaining timers and set them both to T3412 + 4 minutes.
3. Removed the paging check function.

acetcom commented:
@jmasterfunk84 

I made some modifications to your work.

Please let me know if you have any problems.

Thanks you for your contribution.
Sukchan

jmasterfunk84 commented:
Thanks!!

acetcom commented:
@jmasterfunk84 

Based on the standard document below, when the UE is in the IDLE state, you checked the implicit timer and tried to send a message to the UE, but it doesn't work properly.

So, first of all, I deleted the related code.

- TS 24.301 Ch 5.3.7
```
If ISR is not activated, the network behaviour upon expiry of
the mobile reachable timer is network dependent, but typically
the network stops sending paging messages to the UE on the
first expiry, and may take other appropriate actions
```

- TS 24.501 Ch 5.3.7
```
The network behaviour upon expiry of the mobile reachable timer is network dependent,
but typically the network stops sending paging messages to the UE on the first expiry,
and may take other appropriate actions.
```

Please let me know if you have different idea.
Sukchan
